libraries/Microsoft.Bot.Builder.Core.State/DeferredValueStateStorageEntry.cs

```diff
@@ -0,0 +1,39 @@
+﻿namespace Microsoft.Bot.Builder.Core.State
+{
+    public abstract class DeferredValueStateStorageEntry : StateStorageEntry
```

Comment:
The storage provider itself doesn't know the T as it's reading from the underlying representation. It isn't until someone attempts to read the value that the T is known. The storage providers also gain nothing by having more strongly typed instances because, at the end of the day, they all end up doing runtime type inspection to do whatever serialization they need to do anyway.